### PR TITLE
User Feature

### PR DESCRIPTION
**Pull request by ppauljonathan**
Changes made:
  user signup page
  mailer to send verification code to user
  validating verification code

### CODE REVIEW

**Review by shubhamjuneja1 — CHANGES_REQUESTED** [review]

File: db/schema.rb [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+#
+# It's strongly recommended that you check this file into your version control system.
+
+ActiveRecord::Schema[7.0].define(version: 2023_05_10_081504) do
+  create_table "users", force: :cascade do |t|
+    t.string "name", null: false
+    t.string "email", null: false
+    t.string "password_digest", null: false
+    t.integer "credits", default: 0
+    t.boolean "is_admin", default: false
```
[the remark this answers is not in the document]
Comment (ppauljonathan) [reply db/schema.rb:47]: added `enum role: %i[user admin], default: :user`

File: config/routes.rb [hunk 2]
```diff
@@ -1,6 +1,18 @@
 Rails.application.routes.draw do
   # Define your application routes per the DSL in https://guides.rubyonrails.org/routing.html
 
+  resources :users, id: /\d+/, except: %i[new create] do
```
[the remark this answers is not in the document]
Comment (ppauljonathan) [reply config/routes.rb]: i had done it to ensure that `:id` is always an integer, removed now

File: config/routes.rb [hunk 3]
```diff
@@ -1,6 +1,18 @@
...
+    collection do
+      get 'email_verification'
```
[the remark this answers is not in the document]
Comment (ppauljonathan) [reply config/routes.rb]: redone this with devise-like controllers

File: app/controllers/application_controller.rb [hunk 4]
```diff
@@ -1,2 +1,18 @@
 class ApplicationController < ActionController::Base
+  before_action :authorize
+
+  private def redirect_to_homepage_if_logged_in
+    redirect_to users_path if session[:user_id]
+  end
+
+  private def authorize
+    return if session[:user_id]
```
[the remark this answers is not in the document]
Comment (ppauljonathan) [reply app/controllers/application_controller.rb]: done

File: app/controllers/application_controller.rb [hunk 5]
```diff
@@ -1,2 +1,18 @@
...
+    return redirect_to login_url unless cookies.signed[:user_id]
+
+    session[:user_id] = cookies.signed[:user_id]
+  end
+
+  private def current_user
+    @current_user = User.find(session[:user_id])
```
[the remark this answers is not in the document]
Comment (ppauljonathan) [reply app/controllers/application_controller.rb]: done

File: app/controllers/confirmations_controller.rb [hunk 6]
```diff
@@ -0,0 +1,30 @@
+class ConfirmationsController < ApplicationController
+  before_action :redirect_if_logged_in
+  skip_before_action :authorize
+
+  def verify
+    return unless params[:token]
```
[the remark this answers is not in the document]
Comment (ppauljonathan) [reply app/controllers/confirmations_controller.rb:6]: done

File: app/controllers/confirmations_controller.rb [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+    user = User.find_by_verification_token(params[:token])
+    if user.try :verify
```
[the remark this answers is not in the document]
Comment (ppauljonathan) [reply app/controllers/confirmations_controller.rb:9]: done

File: app/controllers/confirmations_controller.rb [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+      redirect_to login_path, notice: 'email was verified successfully, you can log in now'
+    else
+      flash.now[:token_error] = 'Token was invalid, enter email to verify again'
+      render :verify
+    end
+  end
+
+  def resend
+    user = User.find_by_email email_params
+    if user.try :resend_verification_mail
```
[the remark this answers is not in the document]
Comment (ppauljonathan) [reply app/controllers/confirmations_controller.rb:19]: done

File: lib/tasks/admin/new.rake [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+      puts 'Name:'
+      user.name = $stdin.gets.chomp
+      puts 'Email:'
+      user.email = $stdin.gets.chomp
+      puts 'Password:'
+      user.password = $stdin.noecho(&:gets).chomp
+      puts 'Password Confirmation:'
+      user.password_confirmation = $stdin.noecho(&:gets).chomp
+      user.verified_at = Time.now
+      user.is_admin = true
```
[the remark this answers is not in the document]
Comment (ppauljonathan) [reply lib/tasks/admin/new.rake:15]: done

File: app/models/user.rb [hunk 10]
```diff
@@ -0,0 +1,45 @@
+class User < ApplicationRecord
+  include Tokenizable
+
+  validates :name, presence: true
+  validates :email, uniqueness: true, format: { with: URI::MailTo::EMAIL_REGEXP }
```
[the remark this answers is not in the document]
Comment (ppauljonathan) [reply app/models/user.rb:5]: done

File: app/controllers/sessions_controller.rb [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+  def create
+    cookies.signed[:user_id] = { value: @user.id, expires: 24.weeks.from_now } if user_params[:remember_me]
+
+    session[:user_id] = @user.id
+    redirect_to users_path
+  end
+
+  def destroy
+    cookies.delete :user_id
+    session[:user_id] = nil
```
[the remark this answers is not in the document]
Comment (ppauljonathan) [reply app/controllers/sessions_controller.rb:15]: done

File: app/controllers/sessions_controller.rb [hunk 12]
```diff
@@ -0,0 +1,41 @@
+class SessionsController < ApplicationController
+  before_action :redirect_if_logged_in, except: :destroy
+  before_action :set_user_from_email, :authenticate_user, :check_verified_user, only: :create
+  skip_before_action :authorize
+
+  def create
+    cookies.signed[:user_id] = { value: @user.id, expires: 24.weeks.from_now } if user_params[:remember_me]
+
+    session[:user_id] = @user.id
```
[the remark this answers is not in the document]
Comment (ppauljonathan) [reply app/controllers/sessions_controller.rb:9]: done

File: app/controllers/sessions_controller.rb [hunk 13]
```diff
@@ -0,0 +1,41 @@
+class SessionsController < ApplicationController
+  before_action :redirect_if_logged_in, except: :destroy
+  before_action :set_user_from_email, :authenticate_user, :check_verified_user, only: :create
+  skip_before_action :authorize
+
+  def create
+    cookies.signed[:user_id] = { value: @user.id, expires: 24.weeks.from_now } if user_params[:remember_me]
```
[the remark this answers is not in the document]
Comment (ppauljonathan) [reply app/controllers/sessions_controller.rb:7]: done

File: app/controllers/application_controller.rb [hunk 14]
```diff
@@ -1,2 +1,15 @@
...
+  private def redirect_if_logged_in
+    redirect_back_or_to users_path if current_user
+  end
+
+  private def authorize
+    redirect_to login_url unless current_user
+  end
+
+  private def current_user
+    @current_user ||= User.find_by_id(session[:user_id] || cookies.signed[:user_id])
```
[the remark this answers is not in the document]
Comment (ppauljonathan) [reply app/controllers/application_controller.rb:13]: done

File: app/controllers/passwords_controller.rb [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+  def create
+    return unless @user.send_reset_mail
+
+    flash[:notice] = 'Email to reset password has been sent, trigger the action again if not received'
+    redirect_to reset_password_path
+  end
+
+  def edit
+    @user = User.find_by_reset_token params[:token] if params[:token]
+    return if @user
```
[the remark this answers is not in the document]
Comment (ppauljonathan) [reply app/controllers/passwords_controller.rb:15]: changed to before_action

File: app/mailers/user_mailer.rb [hunk 16]
```diff
@@ -0,0 +1,13 @@
+class UserMailer < ApplicationMailer
+
+  # Subject can be set in your I18n file at config/locales/en.yml
+  # with the following lookup:
+  #
+  #   en.user_mailer.verification_email.subject
+  #
+  def verification_email(user_id)
+    @user = User.find(user_id)
```
[the remark this answers is not in the document]
Comment (ppauljonathan) [reply app/mailers/user_mailer.rb:9]: mad before_action

File: app/controllers/passwords_controller.rb [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+
+    redirect_to reset_password_path, notice: 'the token was not valid, please initiate reset password action again'
+  end
+
+  def update
+    @user = User.find(user_params[:id])
+
+    if @user.update(password: user_params[:password])
```
[the remark this answers is not in the document]
Comment (ppauljonathan) [reply app/controllers/passwords_controller.rb:23]: i have made a check box for show password so we may not need confirmation_password

File: lib/tasks/admin/new.rake [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+      user.credits = 5
```
[the remark this answers is not in the document]
Comment (ppauljonathan) [reply lib/tasks/admin/new.rake:16]: done

File: app/models/user.rb [hunk 19]
```diff
@@ -0,0 +1,45 @@
...
+  end
+
+  def verified?
+    verified_at?
+  end
+
+  def resend_verification_mail
+    return false unless update(verification_token: generate_token(:verification), verified_at: nil)
+
+    send_verification_email
```
[the remark this answers is not in the document]
Comment (ppauljonathan) [reply app/models/user.rb:25]: done

File: app/models/user.rb [hunk 20]
```diff
@@ -0,0 +1,45 @@
...
+  def verify
+    update(verified_at: Time.now) unless verified?
+  end
+
+  def verified?
+    verified_at?
+  end
+
+  def resend_verification_mail
+    return false unless update(verification_token: generate_token(:verification), verified_at: nil)
```
[the remark this answers is not in the document]
Comment (ppauljonathan) [reply app/models/user.rb:23]: done

File: app/controllers/users_controller.rb [hunk 21]
```diff
@@ -0,0 +1,34 @@
...
+      render :edit, status: 422
+    end
+  end
+
+  private def set_user
+    @user = User.find(params[:id])
+  end
+
+  private def user_params
+    params.require(:user).permit(:name, :profile_picture)
```
[the remark this answers is not in the document]
Comment (ppauljonathan) [reply app/controllers/users_controller.rb:32]: done in next PR

File: app/controllers/sessions_controller.rb [hunk 22]
```diff
@@ -0,0 +1,41 @@
...
+    params.require(:user).permit(:email, :password, :remember_me)
+  end
+
+  private def set_user_from_email
+    @user = User.find_by(email: user_params[:email])
+    redirect_to login_path, alert: 'Invalid email/password combination' unless @user
+  end
+
+  private def authenticate_user
+    return if @user.try(:authenticate, user_params[:password])
```
Comment (shubhamjuneja1) [inline app/controllers/sessions_controller.rb:31]: do we need to use try here?

**Review by shubhamjuneja1 — APPROVED** [review]

File: app/controllers/sessions_controller.rb (hunk 22 above)
Comment (ppauljonathan) [reply app/controllers/sessions_controller.rb:31]: no need, removed from here

**Pull request merged**